### ISSUE TITLE
Use Auto Embedding for querying but not for indexing

### ISSUE THREAD

**Issue by alihassan0** [typesense/typesense#1254]
## Description

we were trying auto embedding on a new collection but it was using way too much of the server resources and eventually caused the server to crash. 

I was thinking to use the same model for embedding the data before indexing. and then typesense will only need to generate embeddings for the query on runtime. 

I tried multiple solutions to do this yet there doesn't seem to be a way to have the schema use an already generated embeddings and internally generate embeddings for the query. 

## Steps to reproduce

- clone https://github.com/typesense/typesense-instantsearch-semantic-search-demo
- update the schema to auto embed the vectors db. just remove the from field for example.  

## Expected Behavior

you are able to query the collection normally with auto generated embedded for the query. 

## Actual Behavior

the schema will not be accepted and will still need to generate embedding automatically on the server

## Metadata

**Typesense Version**: 0.25.1

**OS**: ubunto

**Comment by alihassan0** [comment]
Thanks a lot for the suggestion .. we tried this and it seemed to work. 

I wonder if there could be an alternative solution to generate the embedding though that doesn't require the to spawn a full typesense server. but I am not sure how to do it exactly !

**Comment by jasonbosco** [comment]
> alternative solution to generate the embedding though that doesn't require the to spawn a full typesense server

You could use a Python script [like this](https://github.com/typesense/showcase-ecommerce-store/blob/33e604a75b81f1de258ff185c775288b27db8335/scripts/vector-generation/main.py) to generate embeddings externally.

**Comment by alihassan0** [comment]
I have tried using this route as well.. but I wasn't able to figure out how to use the same hugging face model in the python script. I have trying using it like this
```python
model = SentenceTransformer('all-MiniLM-L12-v2')
```

but it produces totally different embeddings. any idea how to fix this ?

**Comment by alihassan0** [comment]
Hello @jasonbosco , 

I tried a few more things and I still can't get it product similar embeddings. can you should some light on how to replicate embedding generation similar to typesense auto embeddings but without spawning a typesense instance ?

**Comment by jasonbosco** [comment]
> but it produces totally different embeddings. any idea how to fix this ?

Could you elaborate on this? Any model will generate a set of floating point numbers for vectors, so it would be hard to tell which model was used to generate embeddings, just by looking at the vectors generated.

**Comment by xaptronic** [comment]
> @xaptronic
> 
> If you used Python for generating embeddings but queried using Typesemse, that does not produce totally wrong results? Or they just produce _different_ results (similar but not exact)?

If I create an autoembedding field in typesense, but I provide the embedding vectors (from my python code) when I import the documents into Typesense, then I rely on Typesense autoembedding field to do embedding on the queries, based on some testing, I would say that the results are not _completely wrong_ - the results bring up distantly related document snippets, but are definitely not the same result as when using autoembedding to embed the documents and query.

**Comment by xaptronic** [comment]
I found that the discrepancy for the embedding values between what typesense autoembedding creates vs using sentence transformers from python is that typesense does not perform any sort of normalization after mean_pooling. Not all sentence_transformers have this, but the model I was using does have a [normalization layer](https://huggingface.co/sentence-transformers/all-MiniLM-L12-v2/blob/main/modules.json#L18).

**Comment by kishorenc** [comment]
We have a step that converts both pre-generated and auto-generated embeddings into unit vectors before indexing into HNSW. While this should technically handle all differences, it's not. We will be investigating this in more detail in a few days once we wrap up a few ongoing tasks.

**Comment by xaptronic** [comment]
Here should be a fully reproducible demo of how the same embedding model produces different search results for the same query. This uses dataset I created out of Huggingface, Typesense and Deeplearning.AI documentation. I chunked the text content into max 128 token chunks, then used both `sentence_transformer` and typesense to create embeddings (using the same model) and I exported the typesense embeddings back into the dataset so it imports quickly for testing.

The dataset has two embedding columns: `embedding`, `ts_embedding` - embedding is the set of embeddings created using sentence_transformer, and ts_embedding came from typesense. I built a faiss / pandas dataframe pair as the challenger mechanism.

In this example, different search terms sometimes produce similar results, and other times less similar.
Some examples I tried:
- embedding
- export
- models
- python script
- upload dataset

My assumption is that, since these are using the same model, the results would be identical - there could be a difference in how the similarity is computed in typesense vs faiss, however in various cases it seems the results are entirely wrong.

Hopefully this is helpful.

**Comment by xaptronic** [comment]
I used [this C++ compiled with Typesense bazel](https://gist.github.com/xaptronic/d43e70f3450c86f9a7f809b6bb88c4c6) setup to verify that embedding vectors generated by the ONNX model match the same (unnormalized) embedding vector when using `sentence_transformers`.

**Comment by ozanarmagan** [comment]
@xaptronic Hi, we found a problem with our mean pooling function that causes miscalculation of means of embeddings in some cases, which also leads to difference between embeddings generated by sentence transformers and embeddings generated by our ONNX models. Fix (#1437) will be included to next RC build and upcoming 0.25.2 release.

**Comment by kishorenc** [comment]
@xaptronic 

Can you please check against this build: `0.25.2.rc14`

**Comment by elihoole** [comment]
> ## Description
> we were trying auto embedding on a new collection but it was using way too much of the server resources and eventually caused the server to crash.
> 
> I was thinking to use the same model for embedding the data before indexing. and then typesense will only need to generate embeddings for the query on runtime.
> 
> I tried multiple solutions to do this yet there doesn't seem to be a way to have the schema use an already generated embeddings and internally generate embeddings for the query.
> 
> ## Steps to reproduce
> * clone https://github.com/typesense/typesense-instantsearch-semantic-search-demo
> * update the schema to auto embed the vectors db. just remove the from field for example.
> 
> ## Expected Behavior
> you are able to query the collection normally with auto generated embedded for the query.
> 
> ## Actual Behavior
> the schema will not be accepted and will still need to generate embedding automatically on the server
> 
> ## Metadata
> **Typesense Version**: 0.25.1
> 
> **OS**: ubunto

Hi can you please share the fix. Specifically, did you change anything on the index.js call signatures in the index.js file?

**Comment by alihassan0** [comment]
@elihoole unfortunately I couldn't find a straightforward solution to generating the embeddings before indexing to typesense. all things I tried never led to outputting the same vector array. 

the thing I ended up doing eventually was to index the data to another server other than the production server and export the data with embedding and indexing it then to the production server. that way the production server will no longer need to generate embeddings as it's already in the indexed data
